### PR TITLE
Update `astroid` to 2.12

### LINKED ISSUE #5935: Crash ``argument of type 'Const' is not iterable``
### Bug description

When parsing the following file:

```python
from azure.core import CaseInsensitiveEnumMeta
from enum import Enum
from six import with_metaclass


class PetEnumPy2Metaclass(with_metaclass(CaseInsensitiveEnumMeta, Enum)):
    DOG = "dog"
```

pylint crashed with a ``TypeError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\lint\pylinter.py", line 1034, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\lint\pylinter.py", line 1069, in _check_file
    check_astroid_module(ast_node)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\lint\pylinter.py", line 1203, in check_astroid_module
    retval = self._check_astroid_module(
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\lint\pylinter.py", line 1250, in _check_astroid_module
    walker.walk(node)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\utils\ast_walker.py", line 75, in walk
    self.walk(child)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\utils\ast_walker.py", line 75, in walk
    self.walk(child)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\utils\ast_walker.py", line 75, in walk
    self.walk(child)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\utils\ast_walker.py", line 72, in walk
    callback(astroid)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\checkers\base.py", line 1994, in visit_assignname
    if not list(frame.local_attr_ancestors(node.name)):
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 2606, in local_attr_ancestors
    if name in astroid:
TypeError: argument of type 'Const' is not iterable
```

### Configuration

_No response_

### Command used

```shell
`python -m pylint ..\apistubgentest`
```

### Pylint output

```shell
************* Module apistubgentest.apistubgentest.models._models
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:38:0: C0301: Line too long (110/100) (line-too-long)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:147:0: C0301: Line too long (110/100) (line-too-long)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:1:0: C0114: Missing module docstring (missing-module-docstring)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:17:0: C0115: Missing class docstring (missing-class-docstring)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:18:4: C0203: Metaclass method __getitem__ should have 'cls' as first argument (bad-mcs-method-argument)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:29:0: W0613: Unused argument 'kwargs' (unused-argument)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:29:4: R0201: Method could be a function (no-self-use)
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:25:0: R0903: Too few public methods (1/2) (too-few-public-methods)
Exception on node <AssignName.DOG l.48 at 0x1c284a02770> in file 'F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py'
Traceback (most recent call last):
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\utils\ast_walker.py", line 72, in walk
    callback(astroid)
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\pylint\checkers\base.py", line 1994, in visit_assignname
    if not list(frame.local_attr_ancestors(node.name)):
  File "C:\Users\travi\.pyenv\pyenv-win\versions\3.10.2\lib\site-packages\astroid\nodes\scoped_nodes\scoped_nodes.py", line 2606, in local_attr_ancestors
    if name in astroid:
TypeError: argument of type 'Const' is not iterable
F:\repos\azure-sdk-tools\packages\python-packages\apistubgentest\apistubgentest\models\_models.py:1:0: F0001: Fatal error while checking '..\apistubgentest\apistubgentest\models\_models.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in 'C:\Users\travi\AppData\Local\pylint\pylint\Cache\pylint-crash-2022-03-18-09.txt'. (fatal)

------------------------------------------------------------------
Your code has been rated at 5.68/10 (previous run: 8.50/10, -2.82)
```

### Expected behavior

Not crash

### Pylint version

```shell
Python 3.10.2
pylint 2.12.2
astroid 2.9.3
```

### OS / Environment

Windows 10/11. CMD shell. 

## `six`

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

Fixes #5935
(among other issues to be closed shortly with regression tests in separate PRs, see label "Needs astroid update")

The main change relevant now is `astroid`'s better support for detecting namespace packages.

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
The crashes in the `stdlib` primer are being fixed for 2.12.1, so we'll block this on 2.12.1. I'm about to push a change that will update expected results to fix the windows/pypy failures -- I'm surprised they're not failing on linux/macos -- might have to do with pollution in the environment from a package called "tests": I happened to have one in my environment that had files like this (and once I removed that package, I could produce the test failures just like windows):

```python
from pathlib import Path

from contributors_txt.const import DEFAULT_TEAM_ROLE
from contributors_txt.create_content import Alias, get_aliases

aliases_file = Path(__file__).parent / ".contributors_aliases.json"


def test_basic(recwarn) -> None:
    aliases = get_aliases(aliases_file)
    assert aliases == [
        Alias(
            mails=[
                "66853113+pre-commit-ci[bot]@users.noreply.github.com",
                "49699333+dependabot[bot]@users.noreply.github.com",
            ],
            authoritative_mail="bot@noreply.github.com",
            name="bot",
            team=DEFAULT_TEAM_ROLE,
        ),
        Alias(
            mails=["66853113+pre-commit-ci[bot]@users.noreply.github.com"],
            authoritative_mail="bot@noreply.github.com",
            name="pre-commit-ci[bot]",
            team=DEFAULT_TEAM_ROLE,
        ),
    ]
    assert len(recwarn) == 1
    assert "old copyrite format" in str(recwarn.pop())
```
@Pierre-Sassoulas does that look familiar to you?

**Comment by jacobtylerwalls** [comment]
@Pierre-Sassoulas It's more clear to me now: I think it's a packaging problem with https://github.com/Pierre-Sassoulas/contributors-txt, which installs a `tests` directory in site-packages.

```python
>>> import tests
>>> tests
<module 'tests' from '/Users/.../new/lib/python3.10/site-packages/tests/__init__.py'>
```

The reason we don't have failures on Linux/MacOS is that those install `requirements_test.txt` with `contributors-txt`, and the pypy/windows workflows install `requirements_test_min.txt` without it. I'll push changes to bless the windows/pypy results and then the failures will occur on Linux/MacOS until fixing the packaging issue.

Would you be willing to look into the packaging issue?

**Comment by Pierre-Sassoulas** [comment]
> Would you be willing to look into the packaging issue?

Sure, I'm on it!

File: doc/whatsnew/2/2.15/index.rst [hunk 1]
```diff
@@ -86,6 +89,8 @@ Other Changes
 
   Closes #6953
 
+* Update ``astroid`` to 2.12.0.
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2/2.15/index.rst:92]:
```suggestion
* Update ``astroid`` to 2.12.1.
```

File: pyproject.toml [hunk 2]
```diff
@@ -37,7 +37,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt if you are bumping astroid.
     # Pinned to dev of next minor update to allow editable installs,
     # see https://github.com/PyCQA/astroid/issues/1341
-    "astroid>=2.11.6,<=2.12.0-dev0",
+    "astroid>=2.12.0,<=2.13.0-dev0",
```
Comment (Pierre-Sassoulas) [inline pyproject.toml:40]:
```suggestion
    "astroid>=2.12.1,<=2.13.0-dev0",
```

File: requirements_test_min.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 -e .[testutils,spelling]
-# astroid dependency is also defined in setup.cfg
-astroid==2.11.6  # Pinned to a specific version for tests
+# astroid dependency is also defined in pyproject.toml
+astroid==2.12.0  # Pinned to a specific version for tests
```
Comment (Pierre-Sassoulas) [inline requirements_test_min.txt:3]:
```suggestion
astroid==2.12.1  # Pinned to a specific version for tests
```

**Comment by Pierre-Sassoulas** [comment]
I released a new version of ``contributors-txt`` without ``tests`` directory. But I think that it's probably better if we patch the ``sys.path`` in the current test instead of changing the expected result. (Might be a good thing if we do it in another merge request prior to upgrading astroid and also remove our custom patching function in favor of the current pytest way to do it)

**Comment by jacobtylerwalls** [comment]
I wondered about that. Would you be willing to start that PR?

**Comment by Pierre-Sassoulas** [comment]
Yeap, let me try something :)

**Comment by jacobtylerwalls** [comment]
TODO:
 - [ ] decide if the adjusted test results in 9f67550 and d8d93d6 are tolerable (since `tests` is a package after all)
 - [ ] fix recursion failure with `keras`, see https://github.com/PyCQA/astroid/pull/1660#issuecomment-1179758259

**Comment by AdamWill** [comment]
An interesting thing about the cases where we now get a `tests.` prefix on some strings: this changes depending on *where you run the tests from*. If you run them from outside the top level of the source tree - e.g. go up one level and run something like `pytest pylint/tests`, or do `mkdir src; mv pylint src; pytest` - the `tests.` prefix goes away again (and the tests fail if they've been patched to expect it).

Not sure how much you care about this, but it's there. :D I found it while tinkering about to fix a different problem with running the tests as part of a package build (astroid doesn't like it if you run the tests from the top level of the source tarball but with `PYTHONPATH=/path/to/installroot` to try and run the tests using the 'installed' files, rather than the files from the tarball.

**Comment by jacobtylerwalls** [comment]
> An interesting thing about the cases where we now get a tests. prefix on some strings: this changes depending on where you run the tests from.

Very true. And because that's true, we could probably avoid fiddling with the expected results and just change `cwd` to `tests/data` instead (as some other tests do). Thanks for the inspiration!

@DudeNr33 as a `pyreverse` expert does that sound like a good plan? [Elsewhere](https://github.com/PyCQA/pylint/issues/2763#issuecomment-1145225992) you suggested (for new tests) what I think is a good practice of using dynamically generated files.

**Comment by jacobtylerwalls** [comment]
@AdamWill would you be willing to give this another look?

File: doc/whatsnew/2/2.15/index.rst [hunk 4]
```diff
@@ -86,6 +89,8 @@ Other Changes
 
   Closes #6953
 
+* Update ``astroid`` to 2.12.1.
```
Comment (Pierre-Sassoulas) [inline doc/whatsnew/2/2.15/index.rst:92]:
```suggestion
* Update ``astroid`` to 2.12.
```

File: pyproject.toml [hunk 5]
```diff
@@ -37,7 +37,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt if you are bumping astroid.
     # Pinned to dev of next minor update to allow editable installs,
     # see https://github.com/PyCQA/astroid/issues/1341
-    "astroid>=2.11.6,<=2.12.0-dev0",
+    "astroid>=2.12.1,<=2.13.0-dev0",
```
Comment (Pierre-Sassoulas) [inline pyproject.toml:40]:
```suggestion
    "astroid>=2.12.2,<=2.13.0-dev0",
```

File: requirements_test_min.txt [hunk 6]
```diff
@@ -1,6 +1,6 @@
 -e .[testutils,spelling]
-# astroid dependency is also defined in setup.cfg
-astroid==2.11.6  # Pinned to a specific version for tests
+# astroid dependency is also defined in pyproject.toml
+astroid==2.12.1  # Pinned to a specific version for tests
```
Comment (Pierre-Sassoulas) [inline requirements_test_min.txt:3]:
```suggestion
astroid==2.12.2  # Pinned to a specific version for tests
```

File: pyproject.toml [hunk 7]
```diff
@@ -37,7 +37,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt if you are bumping astroid.
     # Pinned to dev of next minor update to allow editable installs,
     # see https://github.com/PyCQA/astroid/issues/1341
-    "astroid>=2.11.6,<=2.12.0-dev0",
+    "astroid>=2.12.2,<=2.13.0-dev0",
```
Comment (DanielNoord) [inline pyproject.toml:40]:
```suggestion
    "astroid>=2.12.2,<=2.14.0-dev0",
```

File: pyproject.toml [hunk 8]
```diff
@@ -37,7 +37,7 @@ dependencies    = [
     # Also upgrade requirements_test_min.txt if you are bumping astroid.
     # Pinned to dev of next minor update to allow editable installs,
```
Comment (DanielNoord) [inline pyproject.toml:38]:
```suggestion
    # Pinned to dev of second minor update to allow editable installs and fix primer issues,
```

File: doc/user_guide/usage/run.rst [hunk 9]
```diff
@@ -9,22 +9,25 @@ Pylint is meant to be called from the command line. The usage is ::
...
-When ``--recursive=y`` option is used, modules and packages are also accepted as parameters::
+When the ``--recursive=y`` option is used, modules and packages are also accepted as parameters::
 
     pylint --recursive=y mydir mymodule mypackage
 
 This option makes ``pylint`` attempt to discover all modules (files ending with ``.py`` extension)
-and all packages (all directories containing a ``__init__.py`` file).
+and all explicit packages (all directories containing a ``__init__.py`` file).
+As of version 2.15, ``pylint`` will also discover namespace packages but for some edge
+cases that are still under development (where you might still find ``--recursive=y`` useful).
```
Comment (DanielNoord) [inline doc/user_guide/usage/run.rst:28]:
```suggestion
As of version 2.15, ``pylint`` will also discover namespace packages but there are some edge
cases that are still under development (where you might still find ``--recursive=y`` useful).
```

**Comment by DanielNoord** [comment]
We're getting somewhere 🎉 

I "bumped" all caches back to 1 in order not to inflate the number too much.

We also need one fix for some `no-member` warnings which are now recognised due to explicit attributes on `astroid` exceptions.
Rest is just confirmation that changes to the tests are indeed correct.

File: pylint/checkers/imports.py [hunk 10]
```diff
@@ -767,7 +767,9 @@ def _get_imported_module(self, importnode, modname):
                 return None
             self.add_message("relative-beyond-top-level", node=importnode)
         except astroid.AstroidSyntaxError as exc:
-            message = f"Cannot import {modname!r} due to syntax error {str(exc.error)!r}"  # pylint: disable=no-member; false positive
+            message = (
+                f"Cannot import {modname!r} due to syntax error {str(exc.error)!r}"
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/imports.py:771]:
```suggestion
                f"Cannot import {modname} due to syntax error {exc.error}"
```
How about that ?

File: tests/test_self.py [hunk 11]
```diff
@@ -1042,7 +1042,7 @@ def test_regression_parallel_mode_without_filepath(self) -> None:
         path = join(
             HERE, "regrtest_data", "regression_missing_init_3564", "subdirectory/"
         )
-        self._test_output([path, "-j2"], expected_output="No such file or directory")
+        self._test_output([path, "-j2"], expected_output="")
```
Comment (Pierre-Sassoulas) [inline tests/test_self.py:1045]: Isn't this change suspicious ? It seems we might not be testing what we were testing originally.

Comment (jacobtylerwalls) [reply tests/test_self.py:1045]: Good Q. I tried to explain in the commit message, but now we have 23 commits :D

This was added as a regression test for a crash. When written it shouldn't have been depending on the expected output being "we can't lint this namespace package". I think it's still okay to keep the test as a regression test for the crash.

File: pylint/checkers/imports.py (hunk 10 above)
Comment (jacobtylerwalls) [reply pylint/checkers/imports.py:771]: I'm reluctant to merge this in without a test, i.e. we should probably do it separately. What do you think? (Clearly at one point exc.error may not have been a string or could have been arbitrarily set by a user.)

Comment (Pierre-Sassoulas) [reply pylint/checkers/imports.py:771]: Let's do this in a separate MR anyway, there's probably a ton of old %r that were transformed to fstring without refactoring.

File: tests/test_self.py (hunk 11 above)
Comment (Pierre-Sassoulas) [reply tests/test_self.py:1045]: I didn't rewrite the history but we're close to being able to merge now, do you want me to cleanup ?

Comment (Pierre-Sassoulas) [reply tests/test_self.py:1045]: Not sure I saw the actual comment in the commit, it was mixed with a lot of commit from main, I might have removed it by mistake.

File: tests/pyreverse/test_diadefs.py [hunk 12]
```diff
@@ -50,7 +54,8 @@ def HANDLER(default_config: PyreverseConfig) -> DiadefsHandler:
 
 @pytest.fixture(scope="module")
 def PROJECT(get_project):
-    return get_project("data")
+    with _test_cwd(TESTS):
```
Comment (Pierre-Sassoulas) [inline tests/pyreverse/test_diadefs.py:57]: We should use ``tmpdir.as_cwd()`` (until we find something better ?) see https://github.com/PyCQA/pylint/pull/7173

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: tests/messages/func_w0401_package.txt [hunk 13]
```diff
@@ -1,2 +1,2 @@
-R:  1: Cyclic import (input.func_w0401_package.all_the_things -> input.func_w0401_package.thing2)
+R:  1: Cyclic import (tests.input.func_w0401_package.all_the_things -> tests.input.func_w0401_package.thing2)
```
Comment (AdamWill) [inline tests/messages/func_w0401_package.txt:1]: This is the same thing as the pyreverse stuff, I think - should we change to the cwd-based fix here too?

**Comment by DudeNr33** [comment]
> > An interesting thing about the cases where we now get a tests. prefix on some strings: this changes depending on where you run the tests from.
> 
> Very true. And because that's true, we could probably avoid fiddling with the expected results and just change `cwd` to `tests/data` instead (as some other tests do). Thanks for the inspiration!
> 
> @DudeNr33 as a `pyreverse` expert does that sound like a good plan? [Elsewhere](https://github.com/PyCQA/pylint/issues/2763#issuecomment-1145225992) you suggested (for new tests) what I think is a good practice of using dynamically generated files.

Yes, that's fine for me! In reality I'd also expect the user to first `cd` into the root directory for his project.

File: doc/user_guide/usage/run.rst [hunk 14]
```diff
@@ -9,22 +9,25 @@ Pylint is meant to be called from the command line. The usage is ::
...
-When ``--recursive=y`` option is used, modules and packages are also accepted as parameters::
+When the ``--recursive=y`` option is used, modules and packages are also accepted as parameters::
 
     pylint --recursive=y mydir mymodule mypackage
 
 This option makes ``pylint`` attempt to discover all modules (files ending with ``.py`` extension)
-and all packages (all directories containing a ``__init__.py`` file).
+and all explicit packages (all directories containing a ``__init__.py`` file).
+As of version 2.15, ``pylint`` will also discover namespace packages but there are some edge
+cases still under development (where you might still find ``--recursive=y`` useful).
```
Comment (DudeNr33) [inline doc/user_guide/usage/run.rst:28]: I don't quite understand this sentence. From the wording it sounds like it is (in most cases) no longer necessary to use `--recursive=y`, yet the preceding sentences all describe the usage *with* it.

File: tests/pyreverse/test_diadefs.py [hunk 15]
```diff
@@ -100,15 +105,17 @@ class TestDefaultDiadefGenerator:
         ("specialization", "Specialization", "Ancestor"),
     ]
 
-    def test_exctract_relations(
-        self, HANDLER: DiadefsHandler, PROJECT: Project
-    ) -> None:
+    # https://github.com/PyCQA/pylint/issues/2763
+    @pytest.mark.xfail
```
Comment (DudeNr33) [inline tests/pyreverse/test_diadefs.py:109]: I haven't tried this myself yet, but why is the xfail necessary here? 
The linked issue is about working without `__init__.py` files - but the `tests/data` has an `__init__.py` file?

**Review by DudeNr33 — COMMENTED** [review]
I only have an open question why two of the `pyreverse` tests are now marked as xfail - but as we have a lot of other tests for `pyreverse` I think this should not be a blocker, I'm just curious. 😊

File: tests/pyreverse/test_diadefs.py (hunk 12 above)
Comment (jacobtylerwalls) [reply tests/pyreverse/test_diadefs.py:57]: Replied at https://github.com/PyCQA/pylint/pull/7173#issuecomment-1182582877.

File: tests/pyreverse/test_diadefs.py (hunk 15 above)
Comment (jacobtylerwalls) [reply tests/pyreverse/test_diadefs.py:108]: Removed comments in [916e556](https://github.com/PyCQA/pylint/pull/7153/commits/916e556e56c155f151f43e36ca48711b0d10625d)

File: doc/user_guide/usage/run.rst (hunk 14 above)
Comment (jacobtylerwalls) [reply doc/user_guide/usage/run.rst]: Improved comprehensibility in [63d654c](https://github.com/PyCQA/pylint/pull/7153/commits/63d654cf55c911a44d5ecb8ef122c52040f1b483)

File: tests/messages/func_w0401_package.txt (hunk 13 above)
Comment (jacobtylerwalls) [reply tests/messages/func_w0401_package.txt]: cwd-based fix in [6fa6e74](https://github.com/PyCQA/pylint/pull/7153/commits/6fa6e74e465fb145c8d80dabd3a1b0987d6d85b3)

File: tests/pyreverse/test_diadefs.py (hunk 12 above)
Comment (Pierre-Sassoulas) [reply tests/pyreverse/test_diadefs.py:57]: Resolving as we determined that ``_test_cwd`` is actually the best way to do this.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**